colbert/tests/index_updater_test.py

```diff
@@ -0,0 +1,186 @@
+    new_pids = index_updater.add([passage_removed])
+    
+# Search again without reloading the searcher
+    results = searcher.search(question, k=k)
+    top_k_ids_after_append = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids_after_append.append(passage_id)
+        
+    print(f'Top results after appending back the top passage: {top_k_ids_after_append}')
+    
+    if top_k_ids[1:] == top_k_ids_after_append[1:] and top_k_ids_after_append[0] == new_pids[0]:
+        print("REAPPEND SUCCEEDED")
+    else:
+        print("REAPPEND FAILED!!!")
+        all_tests_passed = False
+    
+        
+# Add 4 more copies of the top passages  
+    new_pids.extend(index_updater.add([passage_removed, passage_removed, passage_removed, passage_removed]))
+    
+# Search again without reloading the searcher
+# Now we expect the new 5 identical ids of the top passage to be returned as seearch results
+    results = searcher.search(question, k=k)
+    top_k_ids_after_append = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids_after_append.append(passage_id)
+        
+    print(f'Top results after appending 4 more copies of the top passage: {top_k_ids_after_append}')
+    
+    if set(top_k_ids_after_append) == set(new_pids): # all results should now be the newly appended passages
+        print("REAPPEND 4 more SUCCEEDED")
+    else:
+        print("REAPPEND 4 more FAILED!!!")
+        all_tests_passed = False
+    
+        
+# Persist to disk, reload the searcher and search again
+
+    index_updater.persist_to_disk()
+
+    config = ColBERTConfig(
+            root=experiment_dir,
+            experiment=experiment,
+        )
+    searcher = Searcher(
+            index=experiment,
+            config=config,
+        )
+    results = searcher.search(question, k=k)
+    top_k_ids_reload = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids_reload.append(passage_id)
+        
+# Still we expect the new 5 identical ids of the top passage to be returned as seearch results
+        
+    print(f'Top results after appending 4 more copies of the top passage and reloading: {top_k_ids_reload}')
+    
+    if set(top_k_ids_reload) == set(new_pids): # all results should now be the newly appended passages
```

Comment:
Any plans to add a larger scale, more automatic test?

Any design is fine with me. Here's one option for inspirtation.

- Take a collection with 1M passages (e.g., first 1M passages in MS MARCO). Index it.
- Search for a bunch of queries. Remove a random subset from the top-100 pids (set) for each query.
- Check they're no longer retrieved for the queries.
- Then add back some of these passages, and check they're retrieved.
- Then add back the rest of the deleted passages.
- Then add new passages from MS MARCO that aren't part of the 1M.
- Search with the passages themselves as queries (they'll be truncated; that's fine).
- Check that the passages appear in the top-100 when they are used as queries.

Reply:
Yes, this is a great plan! I'll discuss offline with Keshav today about implementing it (together with tmux and using MS MARCO for the new project as well :)